partner_ref_unique/models/res_partner.py

```diff
@@ -1,38 +1,29 @@
 # Copyright 2016 Antonio Espinosa
+# Copyright 2020 Tecnativa - João Marques
 # License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
 
-from odoo import _, api, fields, models
+from odoo import _, api, models
 from odoo.exceptions import ValidationError
 
 
 class ResPartner(models.Model):
     _inherit = "res.partner"
 
-    # This related is needed in order to trigger the check when changing the
-    # value on res.company
-    partner_ref_unique = fields.Selection(
-        related="company_id.partner_ref_unique", store=True,
-    )
-
-    @api.multi
-    @api.constrains("ref", "is_company", "company_id", "partner_ref_unique")
+    @api.constrains("ref", "is_company", "company_id")
     def _check_ref(self):
-        for partner in self:
-            mode = partner.partner_ref_unique
-            if partner.ref and (
-                mode == "all" or (mode == "companies" and partner.is_company)
-            ):
+        for partner in self.filtered("ref"):
+            # If the company is not defined, adopt default strategy for user company
+            company = partner.company_id or self.env.company
+            mode = company.partner_ref_unique
+            if mode == "all" or (mode == "companies" and partner.is_company):
                 domain = [
                     ("id", "!=", partner.id),
                     ("ref", "=", partner.ref),
                 ]
                 if mode == "companies":
                     domain.append(("is_company", "=", True))
                 other = self.search(domain)
-
-                # active_test is False when called from
```

Comment:
We haven't analyzed this. @Yajo do you remember why this dirty trick for merging. What is the problem merging?

Reply:
I don't remember, but it makes sense that one wants to merge archived partners too.

I think this code is just trying to avoid the validation if you're merging, maybe because during the merge there'll be a valid situation where 2 partners will share the same ref always? :thinking: 

This context doesn't seem to be the most reliable way of doing that, OTOH.

Reply:
That context key was a trick for detecting that you are coming "from a merge operation", so indeed is not correct to check that (you can inject one context key if needed). @joao-p-marques you can check to put one partner with ref, and another without, activate the check, and try to merge one into another (try both ways) to see if there's such problem.

Reply:
Indeed there's a problem when merging that we have already handled. See 9a8c078